packages/flutter/lib/src/cupertino/theme.dart

```diff
@@ -0,0 +1,283 @@
+///
+/// Affects the color and text styles of Cupertino widgets whose styling
+/// are not overridden when constructing the respective widgets instances.
+///
+/// Descendant widgets can retrieve the current [CupertinoThemeData] by calling
+/// [CupertinoTheme.of]. An [InheritedWidget] dependency is created when
+/// an ancestor [CupertinoThemeData] is retrieved via [CupertinoTheme.of].
+///
+/// See also:
+///
+///  * [CupertinoThemeData], specifies the theme's visual styling.
+///  * [CupertinoApp], which will automatically add a [CupertinoTheme].
+///  * [Theme], a Material theme which will automatically add a [CupertinoTheme]
+///    with a [CupertinoThemeData] derived from the Material [ThemeData].
+class CupertinoTheme extends InheritedWidget {
+  /// Creates a [CupertinoTheme] to change descendant Cupertino widgets' styling.
+  ///
+  /// The [data] and [child] parameters must not be null.
+  const CupertinoTheme({
+    Key key,
+    @required this.data,
+    @required Widget child,
+  }) : assert(child != null),
+       assert(data != null),
+       super(key: key, child: child);
+
+  /// The [CupertinoThemeData] styling for this theme.
+  final CupertinoThemeData data;
+
+  @override
+  bool updateShouldNotify(CupertinoTheme oldWidget) => data != oldWidget.data;
+
+  /// Retrieve the [CupertinoThemeData] from an ancestor [CupertinoTheme] widget.
+  ///
+  /// Returns a default [CupertinoThemeData] if no [CupertinoTheme] widgets
+  /// exist in the ancestry tree.
+  static CupertinoThemeData of(BuildContext context) {
+    final CupertinoTheme theme = context.inheritFromWidgetOfExactType(CupertinoTheme);
+    return theme?.data ?? const CupertinoThemeData();
+  }
+}
+
+/// Styling specifications for a [CupertinoTheme].
+///
+/// All constructor parameters can be null, in which case a
+/// [CupertinoColors.activeBlue] based default iOS theme styling is used.
+///
+/// Parameters can also be partially specified, in which case some parameters
+/// will cascade down to other dependent parameters to create a cohesive
+/// visual effect. For instance, if a [primaryColor] is specified, it would
+/// cascade down to affect some fonts in [textTheme] if [textTheme] is not
+/// specified.
+///
+/// See also:
+///
+///  * [CupertinoTheme], in which this [CupertinoThemeData] is inserted.
+///  * [ThemeData], a Material equivalent that also configures Cupertino
+///    styling via a [CupertinoThemeData] subclass [MaterialBasedCupertinoThemeData].
+@immutable
+class CupertinoThemeData extends Diagnosticable {
+  /// Create a [CupertinoTheme] styling specification.
+  ///
+  /// Unspecified parameters default to a reasonable iOS default style.
+  const CupertinoThemeData({
+    Brightness brightness,
+    Color primaryColor,
+    Color primaryContrastingColor,
+    CupertinoTextTheme textTheme,
+    Color barBackgroundColor,
+    Color scaffoldBackgroundColor,
+    Color tableBackgroundColor,
```

Comment:
Can we find better names for `scaffoldBackgroundColor` and `tableBackgroundColor`?

Imagine if the material design implementation defined `fragmentBackgroundColor` and `recyclerViewBackgroundColor` - I think those are the analogs to what we have here.  I think we're leaking artifact information into a color palette.

Can we find a couple design-centric or theme-centric terms for these colors?

Reply:
ya, I do wish I can find some first principle sources for naming these. Unfortunately I don't think Apple themselves derived these colors from some higher order principles (judging by the way how things are named in the Sketch files). So anything we can create are going to be our own conjectures and Apple could break them in the future. 

We're going to have to stick with what we know today. Material is actually going this route too with specific things like SliderThemeData, TabBarTheme, ButtonThemeData etc.

Reply:
Yeah I understand it's not clear what to do.  However, I think we're actually doing harm from an understanding standpoint.  I think we should at least get to neutral on that front.

So for the tableBackgroundColor and scaffoldBackgroundColor, are these colors really going to be limited to those 2 artifacts?  First, which scaffold are we talking about?  CupertinoPageScaffold or CupertinoTabScaffold?  Or both?  If it's both, then are we confident that any future scaffold we introduce will also want this color?  Same for the table.  Is it just for the table, or is it likely that this color is used for listviews?  Or how about as a general content background?

My guess is that both of these colors are actually about some notion of generic content background.  Whether you're displaying a table, a list, a grid, tab content, or something completely custom - this color is probably the main color of the background, no?

If I'm right about that then I think we can find a better name.  If we later re-introduce a table-specific color because we truly need to call out tables on their own, then that's fine.  But I get the impression that at this point the reason you established scaffold and table is because those are the only 2 common artifacts to be found on most Cupertino pages, right?

On a semi-unrelated note, if we're going to introduce CupertinoListView then we should probably avoid any reference to tables.

Reply:
Tweaked it a bit more.

> First, which scaffold are we talking about? CupertinoPageScaffold or CupertinoTabScaffold? Or both? If it's both, then are we confident that any future scaffold we introduce will also want this color?

Both, and looking at the various Sketch files, yes. 

The table background color was a bit special so I removed it for now. You may choose to reintroduce it by privately depending on CupertinoThemeData.brightness and showing CupertinoColors.extraLightBackgroundGray and CupertinoColors.darkBackgroundGray in the CupertinoListView. 

Once we move the various demos to use the CupertinoListView, we can remove things like https://github.com/flutter/flutter/pull/23759/files#diff-05acf49927feb81d25569dde6784c671R717.

Reply:
👍